File: oemkeygen.js
const alphabet = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz-_";
const secret = Buffer.from("StrikeSenseOEMKey2026", "utf8");

const typeAliases = new Map([
  ["24h", "24h"],
  ["24小时", "24h"],
  ["7d", "7d"],
  ["7天", "7d"],
  ["1m", "1m"],
  ["一个月", "1m"],
  ["6m", "6m"],
  ["半年", "6m"],
  ["1y", "1y"],
  ["一年", "1y"],
  ["10y", "10y"],
  ["十年", "10y"],
  ["50y", "50y"],
  ["五十年", "50y"],
  ["forever", "forever"],
  ["永不失效", "forever"],
]);

function normalizeType(kind) {
  const value = typeAliases.get(String(kind).trim());
  if (!value) throw new Error("未知期限类型");
  return value;
}

function fnv1a(text) {
  let h = 2166136261 >>> 0;
  for (const c of Buffer.from(text, "utf8")) {
    h ^= c;
    h = Math.imul(h, 16777619) >>> 0;
  }
  return h >>> 0;
}

function encode64(buf) {
  let out = "";
  let val = 0;
  let bits = -6;
  for (const c of buf) {
    val = (val << 8) + c;
    bits += 8;
    while (bits >= 0) {
      out += alphabet[(val >> bits) & 63];
      bits -= 6;
    }
  }
  if (bits > -6) out += alphabet[((val << 8) >> (bits + 8)) & 63];
  return out;
}

function xorPayload(buf) {
  const out = Buffer.from(buf);
  for (let i = 0; i < out.length; i += 1) {
    out[i] = out[i] ^ secret[i % secret.length] ^ ((i * 29 + 17) & 255);
  }
  return out;
}

function makeKey(stamp, kind) {
  if (!/^\d{8}$/.test(stamp)) throw new Error("时间戳必须是 YYYYMMDD");
  const normalized = normalizeType(kind);
  const body = `SSOEM1|${stamp}|${normalized}`;
  const check = fnv1a(`${body}|StrikeSense`).toString(16);
  return encode64(xorPayload(Buffer.from(`${body}|${check}`, "utf8")));
}

if (typeof module !== "undefined") module.exports = { makeKey, normalizeType };
if (require.main === module) {
  const stamp = process.argv[2] || new Date().toISOString().slice(0, 10).replaceAll("-", "");
  const kind = process.argv[3] || "24小时";
  console.log(makeKey(stamp, kind));
}

File: oemkeygen.py
import argparse
import datetime as dt

ALPHABET = "0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz-_"
SECRET = b"StrikeSenseOEMKey2026"

TYPE_ALIASES = {
    "24h": "24h",
    "24小时": "24h",
    "7d": "7d",
    "7天": "7d",
    "1m": "1m",
    "一个月": "1m",
    "6m": "6m",
    "半年": "6m",
    "1y": "1y",
    "一年": "1y",
    "10y": "10y",
    "十年": "10y",
    "50y": "50y",
    "五十年": "50y",
    "forever": "forever",
    "永不失效": "forever",
}


def normalize_type(kind: str) -> str:
    value = TYPE_ALIASES.get(kind.strip())
    if not value:
        raise ValueError("未知期限类型")
    return value


def fnv1a(text: str) -> int:
    h = 2166136261
    for b in text.encode("utf-8"):
        h ^= b
        h = (h * 16777619) & 0xFFFFFFFF
    return h


def encode64(raw: bytes) -> str:
    out = []
    val = 0
    bits = -6
    for b in raw:
        val = (val << 8) + b
        bits += 8
        while bits >= 0:
            out.append(ALPHABET[(val >> bits) & 63])
            bits -= 6
    if bits > -6:
        out.append(ALPHABET[((val << 8) >> (bits + 8)) & 63])
    return "".join(out)


def xor_payload(raw: bytes) -> bytes:
    data = bytearray(raw)
    for i, b in enumerate(data):
        data[i] = b ^ SECRET[i % len(SECRET)] ^ ((i * 29 + 17) & 255)
    return bytes(data)


def make_key(stamp: str, kind: str) -> str:
    if len(stamp) != 8 or not stamp.isdigit():
        raise ValueError("时间戳必须是 YYYYMMDD")
    normalized = normalize_type(kind)
    body = f"SSOEM1|{stamp}|{normalized}"
    check = format(fnv1a(f"{body}|StrikeSense"), "x")
    return encode64(xor_payload(f"{body}|{check}".encode("utf-8")))


def main() -> None:
    parser = argparse.ArgumentParser(description="StrikeSense OEM key generator")
    parser.add_argument("--stamp", default=dt.date.today().strftime("%Y%m%d"), help="YYYYMMDD, default is today")
    parser.add_argument("--type", default=None, help="24小时/7天/一个月/半年/一年/十年/五十年/永不失效")
    args = parser.parse_args()
    kind = args.type or input("请输入期限类型：").strip()
    print(make_key(args.stamp, kind))


if __name__ == "__main__":
    main()
